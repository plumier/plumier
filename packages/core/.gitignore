!src/generic-controller/filter-grammar.js