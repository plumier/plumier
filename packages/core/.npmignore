test
src
.npmignore
.gitignore
tsconfig.json
package-lock.json
tsconfig.build.json
yarn-error.log
yarn.lock