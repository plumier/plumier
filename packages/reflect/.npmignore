.vscode
coverage
scripts
src 
test
.gitignore
.npmignore
.travis.yml
package-lock.json
tsconfig.build.json
tsconfig.json
yarn-error.log
yarn.lock