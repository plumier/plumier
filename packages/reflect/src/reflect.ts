import { ignore, noop, parameterProperties, type } from "./decorators"
import { createClass, metadata, useCache } from "./helpers"
import { parseFunction } from "./parser"
import { Class, ClassReflection, ObjectReflection, Reflection } from "./types"
import { memberVisitors as v, walkTypeMembersRecursive, walkReflectionMembers } from "./walker"

interface TraverseContext {
    path: any[]
}

type Visitor<Ref, Ctx, Ret> = (r: Ref, c: Ctx) => Ret

function pipe<Ref, Ctx, Ret>(visitors: Visitor<Ref, Ctx, Ret>[]): Visitor<Ref, Ctx, Ret> {
    return (value, ctx) => visitors.reduce((a, b) => !!a ? b(a, ctx) : a, value as any)
}

function reflectClass(target: Class): ClassReflection {
    const ref = walkTypeMembersRecursive(target, {
        target, memberVisitor: pipe([
            // add typescript design types information
            v.addsDesignTypes,
            // add metadata decorators
            v.addsDecorators,
            // add datatype info specified by @type decorator
            v.addsTypeOverridden,
            // add parameter properties
            v.addsParameterProperties,
            // add generic type information
            v.addsGenericOverridden,
            // add typeClassification information
            v.addsTypeClassification,
            // remove @ignore decorator
            v.removeIgnored,
        ]),
        classPath: []
    })
    return walkReflectionMembers(ref, {
        classPath: [], target, parent:ref, memberVisitor: pipe([
            v.addsApplyToDecorator,
            // remove generic type overridden, because it may cause issue for addsTypeOverridden
            v.removeGenericTypeOverridden,
            // add datatype info specified by @type decorator
            v.addsTypeOverridden,
            // add typeClassification information
            v.addsTypeClassification,
        ])})
}

function traverseObject(fn: any, name: string, ctx: TraverseContext): Reflection | undefined {
    if (fn === undefined || fn === null) return
    if (Array.isArray(fn)) return
    if (typeof fn === "object") {
        // CIRCULAR: return immediately
        if (ctx.path.some(x => x === fn)) return
        return reflectObject(fn, name, { path: ctx.path.concat(fn) })
    }
    if (typeof fn === "function" && metadata.isConstructor(fn))
        return reflectClass(fn)
    if (typeof fn === "function")
        return parseFunction(fn)
}

function reflectObject(object: any, name: string, ctx: TraverseContext): ObjectReflection {
    return {
        kind: "Object", name,
        members: Object.keys(object).map(x => traverseObject(object[x], x, ctx)).filter((x): x is Reflection => !!x)
    }
}

function reflectModuleOrClass(opt: string | Class) {
    if (typeof opt === "string") {
        return reflectObject(require(opt), "module", { path: [] })
    }
    else {
        return reflectClass(opt)
    }
}


// --------------------------------------------------------------------- //
// ------------------------------- CACHE ------------------------------- //
// --------------------------------------------------------------------- //

interface ReflectOption {
    /**
     * Flush cached current module/class metadata before processing. 
     */
    flushCache?: true
}

const cacheStore = new Map<string | Class, ClassReflection | ObjectReflection>()
const reflectCached = useCache(cacheStore, reflectModuleOrClass, x => x)

/**
 * Reflect module
 * @param path module name
 */
function reflect(path: string, opt?: Partial<ReflectOption>): ObjectReflection

/**
 * Reflect class
 * @param classType Class 
 */
function reflect(classType: Class, opt?: Partial<ReflectOption>): ClassReflection

function reflect(pathOrClass: string | Class, opt?: Partial<ReflectOption>): ClassReflection | ObjectReflection {
    if (opt?.flushCache)
        cacheStore.delete(pathOrClass)
    return reflectCached(pathOrClass)
}

function flush(pathOrClass?: string | Class) {
    if (pathOrClass)
        cacheStore.delete(pathOrClass)
    else
        cacheStore.clear()
}

// --------------------------------------------------------------------- //
// ----------------------------- DECORATORS ---------------------------- //
// --------------------------------------------------------------------- //

/**
 * Decorator that do nothing, intended to be able to inspect data type
 */
reflect.noop = noop

/**
 * Ignore member from metadata generated
 */
reflect.ignore = ignore

/**
 * Override type definition information. Useful to add type definition for some data type that is erased 
 * after transfile such as Partial<Type> or ReadOnly<Type>
 * 
 * If applied to parameter it will override the parameter type
 * 
 * If applied to property it will override the property type
 * 
 * if applied to method it will overrid the method return value
 * @param type The type overridden
 * @param info Additional information about type (readonly, partial etc)
 */
reflect.type = type

/**
 * Create class dynamically
 */
reflect.create = createClass

/**
 * Mark all constructor parameters as properties
 */
reflect.parameterProperties = parameterProperties

/**
 * Flush metadata cache, call without parameter to clear all
 */
reflect.flush = flush

export { reflect }
