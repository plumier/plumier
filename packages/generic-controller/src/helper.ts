import { RelationDecorator, RelationPropertyDecorator } from "@plumier/core"
import reflect, { Class, generic, GenericTypeDecorator } from "@plumier/reflect"


// --------------------------------------------------------------------- //
// ----------------------- ENTITY RELATION HELPER ---------------------- //
// --------------------------------------------------------------------- //

const genericControllerRegistry = new Map<Class, boolean>()

function updateGenericControllerRegistry(cls: Class) {
    genericControllerRegistry.set(cls, true)
}

function getGenericControllerRelation(controller: Class) {
    const meta = reflect(controller)
    const types = generic.getGenericTypeParameters(controller)
    const parentEntityType = types[0]
    const entityType = types[1]
    const oneToMany = meta.decorators.find((x: RelationPropertyDecorator): x is RelationPropertyDecorator => x.kind === "plumier-meta:relation-prop-name")
    const relation = oneToMany!.name
    return { parentEntityType, entityType, relation, inverseProperty: oneToMany!.inverseProperty }
}

function getGenericControllerInverseProperty(controller: Class) {
    const rel = getGenericControllerRelation(controller)
    return rel.inverseProperty
}

function getGenericControllerOneToOneRelations(type: Class) {
    const meta = reflect(type)
    const result = []
    for (const prop of meta.properties) {
        if (prop.decorators.find((x: RelationDecorator) => x.kind === "plumier-meta:relation") && prop.typeClassification !== "Array") {
            result.push(prop)
        }
    }
    return result
}


export {
    getGenericControllerOneToOneRelations, getGenericControllerInverseProperty,
    getGenericControllerRelation, updateGenericControllerRegistry, genericControllerRegistry
}