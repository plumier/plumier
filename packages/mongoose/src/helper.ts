import { authorize } from "@plumier/core"
import reflect, { Class, useCache } from "@plumier/reflect"

import { RefDecorator } from "./types"

function normalizeEntityNoCache(type: Class) {
    const meta = reflect(type)
    reflect.flush(type)
    for (const prop of meta.properties) {
        const ref = prop.decorators.find((x: RefDecorator) => x.name === "MongooseRef")
        if (ref && prop.typeClassification === "Array") {
            const decorators = [
                authorize.readonly(),
                authorize.writeonly()
            ]
            Reflect.decorate(decorators, type.prototype, prop.name, void 0)
        }
    }
    return { success: true }
}

const normalizeEntityCache = new Map<Class, any>()

export const normalizeEntity = useCache(normalizeEntityCache, normalizeEntityNoCache, x => x)