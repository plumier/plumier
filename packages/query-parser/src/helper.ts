import { AuthorizeDecorator, AuthorizerContext, authPolicy, AuthPolicy, EntityAuthPolicy, executeAuthorizer } from "@plumier/core"
import { Class, generic } from "@plumier/reflect"


const ParserAst = Symbol()

function getDecoratorType(controller: Class, expType: string | Class) {
    // extract generic type from controller if string type provided
    return typeof expType === "string" ? generic.getGenericType(controller, expType) as Class : expType
}

async function isAuthorized(decorators: AuthorizeDecorator[], ctx: AuthorizerContext): Promise<boolean> {
    const authPolicies = ctx.ctx.config.authPolicies.map(x => new x())
    const policies: { decorator: AuthorizeDecorator, isEntityPolicy: boolean }[] = []
    const staticPolicies = []
    // flatten decorator to make each decorator only have one policy
    // and add flag if its an entity policy
    // to make them easily filtered by its flag
    for (const decorator of decorators) {
        for (const policy of decorator.policies) {
            const isEntityPolicy = authPolicies.some(x => x.name === policy && x instanceof EntityAuthPolicy)
            const dec: AuthorizeDecorator = { ...decorator, policies: [policy] }
            if (!isEntityPolicy)
                staticPolicies.push(dec)
            policies.push({ decorator: dec, isEntityPolicy })
        }
    }
    if (policies.every(x => x.isEntityPolicy)) return false
    return executeAuthorizer(staticPolicies, ctx)
}

export { getDecoratorType, ParserAst, isAuthorized }