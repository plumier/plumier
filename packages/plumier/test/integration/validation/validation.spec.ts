import { DefaultDependencyResolver } from "@plumier/core"
import Plumier, {
    AsyncValidatorResult,
    CustomValidator,
    domain,
    RestfulApiFacility,
    route,
    val,
    ValidatorInfo,
    WebApiFacility,
} from "plumier"
import Supertest from "supertest"
import reflect from "tinspector"

import { fixture } from "../../helper"

describe("Validation", () => {
    it("Parameter should be optional by default", async () => {
        class AnimalController {
            get(email: string) { }
        }
        const koa = await fixture(AnimalController).initialize()
        const result = await Supertest(koa.callback())
            .get("/animal/get")
            .expect(200)
        expect(result.body).toMatchSnapshot()
    })

    it("Should validate model with correct path", async () => {
        @domain()
        class AnimalModel {
            constructor(
                public id: number,
                public name: string,
                @val.required()
                public deceased: boolean
            ) { }
        }
        class AnimalController {
            @route.post()
            get(model: AnimalModel) { }
        }
        const koa = await fixture(AnimalController).initialize()
        let result = await Supertest(koa.callback())
            .post("/animal/get")
            .send({ id: "123", name: "Mimi" })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })

    it("Should validate nested model with correct path", async () => {
        @domain()
        class TagModel {
            constructor(public name: string, @val.required() public id: number) { }
        }
        @domain()
        class AnimalModel {
            constructor(
                public id: number,
                public name: string,
                public tag: TagModel
            ) { }
        }
        class AnimalController {
            @route.post()
            get(model: AnimalModel) { }
        }
        const koa = await fixture(AnimalController).initialize()
        let result = await Supertest(koa.callback())
            .post("/animal/get")
            .send({ id: "123", name: "Mimi", tag: { name: "The Tag" } })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })

    it("Should validate parameter", async () => {
        class AnimalController {
            get(@val.email() email: string) { }
        }
        const koa = await fixture(AnimalController).initialize()
        const result = await Supertest(koa.callback())
            .get("/animal/get?email=hello")
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })

    it("Should skip validation if no query provided", async () => {
        class AnimalController {
            get(@val.email() email: string) { }
        }
        const koa = await fixture(AnimalController).initialize()
        const result = await Supertest(koa.callback())
            .get("/animal/get")
            .expect(200)
    })

    it("Should validate parameter properly", async () => {
        class AnimalController {
            get(@val.email() email: string) { }
        }
        const koa = await fixture(AnimalController).initialize()
        const result = await Supertest(koa.callback())
            .get("/animal/get?email=m.ketut@gmail.com")
            .expect(200)
    })

    it("Should optionally check on partial validation", async () => {
        @domain()
        class AnimalModel {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean
            ) { }
        }
        class AnimalController {
            @route.post()
            get(@val.partial(AnimalModel) model: Partial<AnimalModel>) {
                expect(typeof model.id).toBe("number")
                expect(typeof model.deceased).toBe("boolean")
            }
        }
        const koa = await fixture(AnimalController).initialize()
        let result = await Supertest(koa.callback())
            .post("/animal/get")
            .send({ id: "123", deceased: "True" })
            .expect(200)
    })

})

describe("Error handling", () => {
    function customValidator() {
        return val.custom(x => { throw new Error("ERROR") })
    }
    it("Should provide correct information when error inside custom validator", async () => {
        const hook = jest.fn()
        class AnimalController {
            get(@customValidator() email: string) { }
        }
        const koa = await fixture(AnimalController).initialize()
        koa.on("error", e => {
            hook(e)
        })
        const result = await Supertest(koa.callback())
            .get("/animal/get?email=hello")
            .expect(500)
        expect(hook.mock.calls[0][0].stack).toContain("validation.spec")
    })
})

describe("Decouple Validation Logic", () => {
    @domain()
    class Person {
        constructor(
            @val.custom("18+only")
            public age: number
        ) { }
    }
    class PersonController {
        @route.post()
        save(data: Person) { }
    }
    const resolver = new DefaultDependencyResolver()

    @resolver.register("18+only")
    class AgeValidator implements CustomValidator {
        validate(value: any) {
            if (parseInt(value) <= 18)
                return "Only 18+ allowed"
        }
    }


    it("Should validate using decouple logic from setting", async () => {
        const koa = await fixture(PersonController, { dependencyResolver: resolver }).initialize()
        const result = await Supertest(koa.callback())
            .post("/person/save")
            .send({ age: 9 })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })

    it("Should validate using decouple logic from WebApiFacility", async () => {
        const koa = await new Plumier()
            .set(new WebApiFacility({ dependencyResolver: resolver, controller: PersonController }))
            .set({ mode: "production" })
            .initialize()
        const result = await Supertest(koa.callback())
            .post("/person/save")
            .send({ age: 9 })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })

    it("Should validate using decouple logic from RestfulApiFacility", async () => {
        const koa = await new Plumier()
            .set(new RestfulApiFacility({ dependencyResolver: resolver, controller: PersonController }))
            .set({ mode: "production" })
            .initialize()
        const result = await Supertest(koa.callback())
            .post("/person/save")
            .send({ age: 9 })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })
})

describe("Custom Validation", () => {

    it("Should provided correct information for custom validation", async () => {
        async function customValidator(val: any, info: ValidatorInfo) {
            expect(info.name).toBe("data")
            expect(info.parent).toBeUndefined()
            expect(info.ctx.route).toMatchSnapshot()
            return undefined
        }
        class UserController {
            @route.post()
            save(@val.custom(customValidator) data: string) { }
        }
        const koa = await fixture(UserController).initialize()
        await Supertest(koa.callback())
            .post("/user/save")
            .send({ data: "abc" })
            .expect(200)
    })

    it("Should able to use sync function as custom validator", async () => {
        class UserController {
            @route.post()
            save(@val.custom(val => val < 18 ? "Must greater than 18" : undefined) data: number) { }
        }
        const koa = await fixture(UserController).initialize()
        const { body } = await Supertest(koa.callback())
            .post("/user/save")
            .send({ data: 12 })
            .expect(422)
        expect(body).toMatchSnapshot()
    })

    it("Should able to use class based custom validator", async () => {
        class AgeValidator implements CustomValidator {
            validate(val: any) {
                if (val < 18)
                    return "Must greater than 18"
            }
        }
        class UserController {
            @route.post()
            save(@val.custom(new AgeValidator()) data: number) { }
        }
        const koa = await fixture(UserController).initialize()
        const { body } = await Supertest(koa.callback())
            .post("/user/save")
            .send({ data: 12 })
            .expect(422)
        expect(body).toMatchSnapshot()
    })

    it("Should able to return AsyncValidatorResult from sync validation", async () => {
        class UserController {
            @route.post()
            save(@val.custom(v => v < 18 ? val.result("other", "Must greater than 18") : undefined) data: number) { }
        }
        const koa = await fixture(UserController).initialize()
        const { body } = await Supertest(koa.callback())
            .post("/user/save")
            .send({ data: 12 })
            .expect(422)
        expect(body).toMatchSnapshot()
    })

    it("Should able to return AsyncValidatorResult with multiple messages", async () => {
        class UserController {
            @route.post()
            save(@val.custom(v => v < 18 ? val.result("other", ["Must greater", "Than 18"]) : undefined) data: number) { }
        }
        const koa = await fixture(UserController).initialize()
        const { body } = await Supertest(koa.callback())
            .post("/user/save")
            .send({ data: 12 })
            .expect(422)
        expect(body).toMatchSnapshot()
    })


    it("Should able to use async function as custom validator", async () => {
        class UserController {
            @route.post()
            save(@val.custom(async val => val < 18 ? "Must greater than 18" : undefined) data: number) { }
        }
        const koa = await fixture(UserController).initialize()
        const { body } = await Supertest(koa.callback())
            .post("/user/save")
            .send({ data: 12 })
            .expect(422)
        expect(body).toMatchSnapshot()
    })

    it("Should provide parent value information", async () => {
        @domain()
        class ClientModel {
            constructor(
                public password: string,
                @val.custom(async (val, info) => {
                    const pwd = (info.parent!.value as ClientModel).password
                    return val !== pwd ? "Password doesn't match" : undefined
                })
                public confirmPassword: string
            ) { }
        }
        class UserController {
            @route.post()
            save(data: ClientModel) { }
        }

        const koa = await fixture(UserController).initialize()
        const result = await Supertest(koa.callback())
            .post("/user/save")
            .send({ password: "abcde", confirmPassword: "efghi" })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })

    it("Should able to combine TypedConverter validator and custom validator", async () => {
        function longerThanTwenty() {
            return val.custom(async x => x.length > 20 ? undefined : "String must be longer than 20")
        }
        @reflect.parameterProperties()
        class EmailOnly {
            constructor(
                @val.email()
                @longerThanTwenty()
                public email: string
            ) { }
        }
        class UserController {
            @route.post()
            save(data: EmailOnly) { }
        }

        const koa = await fixture(UserController).initialize()
        const result = await Supertest(koa.callback())
            .post("/user/save")
            .send({ email: "lorem ipsum" })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })

    it("Should validate using decouple logic", async () => {
        const registry = new DefaultDependencyResolver()

        @registry.register("18+only")
        class AgeValidator implements CustomValidator {
            validate(value: any, info: ValidatorInfo): string | AsyncValidatorResult[] | Promise<string | AsyncValidatorResult[] | undefined> | undefined {
                if (parseInt(value) <= 18)
                    return "Only 18+ allowed"
            }
        }

        @reflect.parameterProperties()
        class EmailOnly {
            constructor(
                @val.custom("18+only")
                public age: number
            ) { }
        }

        class UserController {
            @route.post()
            save(data: EmailOnly) { }
        }

        const koa = await fixture(UserController)
            .set({ dependencyResolver: registry })
            .initialize()
        const result = await Supertest(koa.callback())
            .post("/user/save")
            .send({ age: "12" })
            .expect(422)
        await Supertest(koa.callback())
            .post("/user/save")
            .send({ age: "20" })
            .expect(200)
        expect(result.body).toMatchSnapshot()
    })

    it("Should throw proper error if no validator store provided", async () => {
        function only18Plus() {
            return val.custom("18+only")
        }
        @reflect.parameterProperties()
        class EmailOnly {
            constructor(
                @only18Plus()
                public age: number
            ) { }
        }
        class UserController {
            @route.post()
            save(data: EmailOnly) { }
        }

        const koa = await fixture(UserController).initialize()
        koa.on("error", () => { })
        await Supertest(koa.callback())
            .post("/user/save")
            .send({ age: "12" })
            .expect(500)
    })

    it("Should validate using custom validation", async () => {
        @domain()
        class ClientModel {
            constructor(
                public password: string,
                @val.custom(async (val, info) => {
                    const pwd = (info.ctx.request.body as ClientModel).password
                    return val !== pwd ? "Password doesn't match" : undefined
                })
                public confirmPassword: string
            ) { }
        }
        class UserController {
            @route.post()
            save(data: ClientModel) { }
        }

        const koa = await fixture(UserController).initialize()
        const result = await Supertest(koa.callback())
            .post("/user/save")
            .send({ password: "abcde", confirmPassword: "efghi" })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })

    it("Should be able to validate class and return several validation result", async () => {
        function checkConfirmPassword() {
            return val.custom( x => {
                if(x.password !== x.confirmPassword)
                    return val.result("confirmPassword", "Password is not the same")
            })
        }
        @domain()
        class User {
            constructor(
                public password: string,
                public confirmPassword: string
            ) { }
        }
        class UsersController {
            @route.post()
            get(@checkConfirmPassword() model: User) { }
        }
        const koa = await fixture(UsersController).initialize()
        let result = await Supertest(koa.callback())
            .post("/users/get")
            .send({ password: "111111", confirmPassword: "2222222" })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })

    it("Should be able to validate class from the class decorator", async () => {
        function checkConfirmPassword() {
            return val.custom( x => {
                if(x.password !== x.confirmPassword)
                    return val.result("confirmPassword", "Password is not the same")
            })
        }
        @domain()
        @checkConfirmPassword()
        class User {
            constructor(
                public password: string,
                public confirmPassword: string
            ) { }
        }
        class UsersController {
            @route.post()
            get(model: User) { }
        }
        const koa = await fixture(UsersController).initialize()
        let result = await Supertest(koa.callback())
            .post("/users/get")
            .send({ password: "111111", confirmPassword: "2222222" })
            .expect(422)
        expect(result.body).toMatchSnapshot()
    })
})