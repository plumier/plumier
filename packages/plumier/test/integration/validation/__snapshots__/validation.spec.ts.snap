// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Validation Should able to combine TypedConverter validator and custom validator 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Invalid email address",
        "String must be longer than 20",
      ],
      "path": Array [
        "data",
        "email",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Custom Validation Should provided correct information for custom validation 1`] = `
Object {
  "action": Object {
    "decorators": Array [
      Object {
        "method": "post",
        "name": "Route",
        "url": undefined,
      },
    ],
    "kind": "Method",
    "name": "save",
    "parameters": Array [
      Object {
        "decorators": Array [
          Object {
            "type": "ValidatorDecorator",
            "validator": customValidator,
          },
        ],
        "kind": "Parameter",
        "name": "data",
        "type": String,
        "typeClassification": "Primitive",
      },
    ],
    "returnType": undefined,
    "typeClassification": undefined,
  },
  "controller": Object {
    "ctor": Object {
      "kind": "Constructor",
      "name": "constructor",
      "parameters": Array [],
    },
    "decorators": Array [],
    "kind": "Class",
    "methods": Array [
      Object {
        "decorators": Array [
          Object {
            "method": "post",
            "name": "Route",
            "url": undefined,
          },
        ],
        "kind": "Method",
        "name": "save",
        "parameters": Array [
          Object {
            "decorators": Array [
              Object {
                "type": "ValidatorDecorator",
                "validator": customValidator,
              },
            ],
            "kind": "Parameter",
            "name": "data",
            "type": String,
            "typeClassification": "Primitive",
          },
        ],
        "returnType": undefined,
        "typeClassification": undefined,
      },
    ],
    "name": "UserController",
    "properties": Array [],
    "type": UserController,
    "typeClassification": "Class",
  },
  "method": "post",
  "url": "/user/save",
}
`;

exports[`Custom Validation Should validate using custom validation 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Password doesn't match",
      ],
      "path": Array [
        "data",
        "confirmPassword",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Custom Validation Should validate using decouple logic 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Only 18+ allowed",
      ],
      "path": Array [
        "data",
        "age",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Decouple Validation Logic Should validate using decouple logic from RestfulApiFacility 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Only 18+ allowed",
      ],
      "path": Array [
        "data",
        "age",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Decouple Validation Logic Should validate using decouple logic from WebApiFacility 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Only 18+ allowed",
      ],
      "path": Array [
        "data",
        "age",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Decouple Validation Logic Should validate using decouple logic from setting 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Only 18+ allowed",
      ],
      "path": Array [
        "data",
        "age",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Validation Parameter should be mandatory by default 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": Array [
        "email",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Validation Should be able to validate class and return several validation result 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Password is not the same",
      ],
      "path": Array [
        "model",
        "confirmPassword",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Validation Should validate model with correct path 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": Array [
        "model",
        "deceased",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Validation Should validate nested model with correct path 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": Array [
        "model",
        "tag",
        "id",
      ],
    },
  ],
  "status": 422,
}
`;

exports[`Validation Should validate parameter 1`] = `
Object {
  "message": Array [
    Object {
      "messages": Array [
        "Invalid email address",
      ],
      "path": Array [
        "email",
      ],
    },
  ],
  "status": 422,
}
`;
